runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp

```diff
@@ -323,8 +323,172 @@ J9::ARM64::TreeEvaluator::DIVCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
       snippet->gcMap().setGCRegisterMask(0xffffffff);
       }
 
-   cg->evaluate(node->getFirstChild());
+TR::Register *
+J9::ARM64::TreeEvaluator::ArrayCHKEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   TR::Register *obj1Reg = cg->evaluate(node->getFirstChild());
+   TR::Register *obj2Reg = cg->evaluate(node->getSecondChild());
+   TR::Register *tmp1Reg = cg->allocateRegister();
+   TR::Register *tmp2Reg = cg->allocateRegister();
+   TR::LabelSymbol *doneLabel = generateLabelSymbol(cg);
+   TR::RegisterDependencyConditions *deps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(4, 4, cg->trMemory());
+   TR::Compilation* comp = cg->comp();
+   TR::LabelSymbol *snippetLabel;
+   TR::Instruction *gcPoint;
+   TR::Snippet *snippet;
+   bool isLong = TR::Compiler->target.is64Bit();
+
+   TR::addDependency(deps, obj1Reg, TR::RealRegister::NoReg, TR_GPR, cg);
+   TR::addDependency(deps, obj2Reg, TR::RealRegister::NoReg, TR_GPR, cg);
+   TR::addDependency(deps, tmp1Reg, TR::RealRegister::NoReg, TR_GPR, cg);
+   TR::addDependency(deps, tmp2Reg, TR::RealRegister::NoReg, TR_GPR, cg);
+
+   // Same array, we are done.
+   generateCompareInstruction(cg, node, obj1Reg, obj2Reg);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, doneLabel, TR::CC_EQ);
+
+   // If we know nothing about either object, test object1 first. It has to be an array.
+   if ((!node->isArrayChkPrimitiveArray1())
+         && (!node->isArrayChkReferenceArray1())
+         && (!node->isArrayChkPrimitiveArray2())
+         && (!node->isArrayChkReferenceArray2()))
+      {
+#ifdef OMR_GC_COMPRESSED_POINTERS
+      generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, tmp1Reg,
+            new (cg->trHeapMemory()) TR::MemoryReference(obj1Reg, (int32_t) TR::Compiler->om.offsetOfObjectVftField(), cg));
+#else
+      generateTrg1MemInstruction(cg, isLong ? TR::InstOpCode::ldrx : TR::InstOpCode::ldrw, node, tmp1Reg,
+            new (cg->trHeapMemory()) TR::MemoryReference(obj1Reg, TR::Compiler->om.offsetOfObjectVftField(), cg));
+#endif
+      TR::TreeEvaluator::generateVFTMaskInstruction(cg, node, tmp1Reg);
+      generateTrg1MemInstruction(cg, TR::InstOpCode::ldrw, node, tmp1Reg,
+            new (cg->trHeapMemory()) TR::MemoryReference(tmp1Reg, offsetof(J9Class, classDepthAndFlags), cg));
+
+      loadConstant32(cg, node, J9AccClassRAMArray, tmp2Reg);
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::andw, node, tmp2Reg, tmp1Reg, tmp2Reg);
+
+      generateCompareImmInstruction(cg, node, tmp2Reg, NULLVALUE);
+
+      snippetLabel = generateLabelSymbol(cg);
+      gcPoint = generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, snippetLabel, TR::CC_EQ);
```

Comment:
You can use `generateCompareBranchInstruction(TR::InstOpCode::cbzx)` for comparing with zero to branch, replacing `generateCompareImmInstruction()` and `generateConditionalBranchInstruction()`.